rfcs/0154-Migrate-taskcluster-to-postgres.md

```diff
@@ -0,0 +1,161 @@
+    * What tasks are pending right now?
+    * How many tasks are [actually pending and not cancelled](https://bugzilla.mozilla.org/show_bug.cgi?id=1434851)?
+  * Integration of task-related and worker-related status.
+    * The approach we've taken with Azure doesn't allow tables to read from
+      other tables.
+    * What tasks has this worker performed?
+  * Ad-hoc analytics.
+    * How much compute time do we use per push?
+    * What do we pay for wpt?
+  * More sophisticated dependency resolution [in the case of task failure](https://bugzilla.mozilla.org/show_bug.cgi?id=1443503#c8)
+  * Dynamically change task priorities.
+
+# Details
+
+## Transition
+
+The transition to Postgres will be done in two steps:
+1. Transfer the data from Azure to Postgres without data loss. This will not aim
+   at normalizing the data but rather prioritize a safe and easy transition.  In
+   other words, the data will be backed by Postgres, but still behaving like
+   Azure.
+2. Go fully into normalized SQL tables.
+
+In both steps, Taskcluster will get confidence in the correctness and
+performance of the implementation by load testing it on the development and
+staging environments prior to running this on the production environment.
+
+Upon completion of step 1, deployers of Taskcluster will need to provide a DB
+URL (maybe several) then run a script to "load" data from Azure to Postgres
+during a downtime.
+
+There will be a version of Taskcluster that supports Postgres and not Azure,
+with the previous version supporting Azure and not Postgres. So upgrading to
+that version is something that happens in a downtime.
+
+The point of no return will be 2 weeks. This should give us enough time to see
+if anything goes wrong. At any time before the point of no return, rolling back
+will consist of switching back to a previous release of Taskcluster and
+reverting any kubernetes config changes that were done for this migration. The
+impact of rolling back will be data loss of all data after the planned downtime.
+There will be no data loss if rolling back happens during the downtime.
+
+### Tracing
+
+Taskcluster will use New Relic to a have better visibility of the database,
+especially before the point of no return (2 weeks). This should help in
+monitoring performance and easier troubleshooting.
+
+### Load Testing
+
+Taskcluster will get confidence in the correctness and performance of its
+implementation with load testing the database to simulate a realistic user load
+prior to using postgres in production. We will load test the data volume in the
+database as well as the number of queries coming in.
+
+#### Data Volume
+
+To test the scalability and performance of the system, we will do an import of
+the FirefoxCI production database (minus the secrets) into the postgres database
+on the staging deployment and then observe if the database crashes or if there
+are any noticeable performance issues that arise.
```

Comment:
Great, thanks for adding this! Let's make sure we do the parallel request testing on the system to which we've imported production amounts of data. We want to test 2 things: 
1) Can production quantities of data be imported successfully?
2) Does the DB perform as expected when it has production quantities of data in it?
As currently written, this tests 1 but not 2. To test both, we could just make sure to do the Parallel Requests testing on the same instance that we've imported the prod quantities of data into.

Reply:
Good observation. I'll update the RFC to make this clearer.

Reply:
Done.